File: backend/schemas/post.py
from pydantic import BaseModel
from typing import List, Optional, Dict

# 共用
class PostImageItem(BaseModel):
    image_id: str
    url: str
    width: Optional[int] = None
    height: Optional[int] = None
    order: int

class PostListItem(BaseModel):
    post_id: str
    content: str
    images: List[PostImageItem]
    is_liked: bool
    like_count: int
    comment_count: int
    created_at: str
    user_id: str
    username: str

class PostListPagination(BaseModel):
    limit: int
    cursor_post_id: Optional[str] = None
    total: int  # 總頁數

class PostListData(BaseModel):
    posts: List[PostListItem]
    pagination: PostListPagination

class PostListResponse(BaseModel):
    data: PostListData

class PostDetailCommentUser(BaseModel):
    user_id: str
    name: Optional[str] = None
    username: str
    metadata: Optional[Dict] = {}

class PostDetailComment(BaseModel):
    comment_id: str
    content: str
    created_at: str
    user: PostDetailCommentUser

class PostDetailData(BaseModel):
    post_id: str
    content: str
    images: List[PostImageItem]
    comments: List[PostDetailComment]
    created_at: str
    is_liked: bool
    like_count: int
    comment_count: int
    user_id: str
    username: str

class PostDetailResponse(BaseModel):
    data: PostDetailData
    message: str = "ok"

class PostCreateData(BaseModel):
    post_id: str

class PostCreateResponse(BaseModel):
    data: PostCreateData
    message: str = "ok"

class PostUpdateResponse(PostCreateResponse):
    pass

class PostDeleteResponse(PostCreateResponse):
    pass

class CommentsPagination(BaseModel):
    page: int
    limit: int
    total: int  # 總頁數

class CommentListData(BaseModel):
    comments: List[PostDetailComment]
    pagination: CommentsPagination

class CommentListResponse(BaseModel):
    data: CommentListData
    message: str = "ok"

class CommentCreateData(BaseModel):
    comment_id: str

class CommentCreateResponse(BaseModel):
    data: CommentCreateData
    message: str = "ok"

class CommentUpdateResponse(CommentCreateResponse):
    pass

class CommentDeleteResponse(CommentCreateResponse):
    pass

File: frontend/src/types/post.ts
// ==========================================
// Type definitions corresponding to backend schemas/post.py
// ==========================================

// Corresponds to PostImageItem
export interface PostImage {
  image_id: string;
  url: string;
  width?: number;
  height?: number;
  order: number;
}

// Corresponds to PostListItem (post in list)
export interface PostListItem {
  post_id: string;
  content: string;
  images: PostImage[];
  is_liked: boolean;
  like_count: number;
  comment_count: number;
  created_at: string;
  user_id: string;
  username: string;
}

// Corresponds to PostListPagination
export interface PostListPagination {
  limit: number;
  cursor_post_id: string | null;
  total: number;
}

// Corresponds to PostListData
export interface PostListData {
  posts: PostListItem[];
  pagination: PostListPagination;
}

// Corresponds to PostListResponse (full response from GET /api/posts)
export interface PostListResponse {
  data: PostListData;
}

// Corresponds to PostDetailCommentUser
export interface PostDetailCommentUser {
  user_id: string;
  name?: string;
  metadata?: Record<string, any>;
}

// Corresponds to PostDetailComment
export interface PostDetailComment {
  comment_id: string;
  content: string;
  created_at: string;
  user: PostDetailCommentUser;
}

// Corresponds to PostDetailData (single post detail)
export interface PostDetailData {
  post_id: string;
  content: string;
  images: PostImage[];
  comments: PostDetailComment[];
  created_at: string;
  is_liked: boolean;
  like_count: number;
  comment_count: number;
  user_id: string;
  username: string;
}

// Corresponds to PostDetailResponse (full response from GET /api/posts/{post_id})
export interface PostDetailResponse {
  data: PostDetailData;
  message: string;
}

// Corresponds to PostCreateData
export interface PostCreateData {
  post_id: string;
}

// Corresponds to PostCreateResponse (full response from POST /api/posts/)
export interface PostCreateResponse {
  data: PostCreateData;
  message: string;
}

// Corresponds to PostUpdateResponse and PostDeleteResponse
export interface PostUpdateResponse extends PostCreateResponse {}
export interface PostDeleteResponse extends PostCreateResponse {}

// ==========================================
// Transformed types for frontend UI (convenient for frontend use)
// ==========================================

/**
 * Transformed post type for frontend UI display
 * Converts backend snake_case to frontend camelCase
 */
export interface Post {
  id: string;                // converted from post_id
  content: string;
  images: PostImage[];
  isLiked: boolean;          // converted from is_liked
  likeCount: number;         // converted from like_count
  commentCount: number;      // converted from comment_count
  createdAt: string;         // converted from created_at
  userId: string;            // converted from user_id
  username: string;          // converted from username
}

/**
 * Transformed post detail type (includes comments)
 */
export interface PostDetail extends Post {
  comments: PostDetailComment[];
}

// ==========================================
// Frontend input types
// ==========================================

/**
 * Input for creating post (corresponds to backend Form data)
 * Backend expects: content (Form) + images (File)
 */
export interface CreatePostInput {
  content: string;           // backend only needs content
  images?: File[];           // optional image uploads
}

/**
 * Input for updating post
 */
export interface UpdatePostInput {
  content?: string;
}
